docs/infrastructure/logging.md

```diff
@@ -257,6 +264,116 @@ determines how often logger should send its collected events to sentry.io
+[org.enso.logging.service.logback.DeferredProcessingSocketAppender](https://github.com/enso-org/enso/blob/566d3d503e35ccf8c363facd590d1332683bce3a/lib/scala/logging-service-logback/src/main/java/org/enso/logging/service/logback/DeferredProcessingSocketAppender.java)
+to send log events to the logging server. The logging server then properly
+dispatches the received logging event to all the appenders in
+[SocketLoggingNode](https://github.com/enso-org/enso/blob/3e0b4dd7413e45373dc4ac26f124ee6aedffd50c/lib/scala/logging-service-logback/src/main/java/org/enso/logging/service/logback/SocketLoggingNode.java#L87-L94).
+
+## Telemetry
+
+Collection of telemetry events is implemented via logging. The motivation for
+that is that we want to be able to collect telemetry from all possible sources -
+engine, standard libraries, language server and project manager. All the
+telemetry events are visible in our OpenSearch dashboard. To properly send
+telemetry logs to the cloud, the used logger and the log message must conform to
+the specification described below. The original proposal of the specification is
+in
+[#12464](https://github.com/enso-org/enso/issues/12464#issuecomment-2713545076).
+
+### Logger namespace
+
+To send telemetry data to our cloud endpoint, a logger inside
+`org.enso.telemetry` namespace must be used. To create such a logger from Java,
+use something like this:
+
+```java
+var logger = org.slf4j.LoggerFactory.getLogger("org.enso.telemetry.MyLogger");
+```
+
+To create this logger in Enso, use:
+
+```
+polyglot java import org.slf4j.LoggerFactory
+logger = LoggerFactory.getLogger "org.enso.telemetry.MyLogger"
+```
+
+### LogEvent format
+
+The format of the message passed to the logger is important. If it does not
+follow this specification, it will not be forwarded to the cloud endpoint. The
+message consist of two parts: `message` and `arguments` delimited by a colon:
+`<message>: <arguments>`. The `message` part is a string that describes the
+event. The `arguments` part is a mapping of `name={}` pairs separated by a
+comma. Note that for the compatibility with standard `ch.qos.logback` appenders,
+it is important that the value of argument is `{}`. The actual argument values
+are passed to the message as argument array via
+`ILoggingEvent.getArgumentArray()` method. Number of these arguments must match
+number of the arguments given in the message.
+
+An example in Java is:
+
+```java
+logger.trace("MyEvent: name={}, age={}", "John", 42);
+```
+
+In Enso, one has to use
+[org.slf4j.spi.LoggingEventBuilder](https://github.com/qos-ch/slf4j/blob/master/slf4j-api/src/main/java/org/slf4j/spi/LoggingEventBuilder.java)
+like this:
+
+```
+event_bldr = logger.atTrace
+event_bldr.setMessage "MyEvent: name={}, age={}"
+event_bldr.addArgument "John"
+event_bldr.addArgument 42
+event_bldr.log
+```
+
+The reason for that is that the `trace` method accepts vararg and there is no
+way to pass a vararg arguments from Enso to Java.
```

Comment:
This is a Truffle API limitation. Not something that could be fixed on Enso side.

Reply:
Note added in 6ae6a7a07ee7c2b05008229269c6da070531f102